### PR TITLE
Support almost ALL types in/out. (e.x. Enum, OffsetDateTime, BigDecimal, double, Float... )

### PR DESCRIPTION
**Pull request by QianJin2013**
### [Java-Client-Driver] There are many data type can not be store to or retrieve from db just because the type conversion logic have some problems or not enough.

These problematic types are (23 types):
```
enum
BigDecimal, BigInteger
Boolean, boolean
Double, double, Float, float, Long, long, Integer, int, Short, short, Byte, byte
OffsetDateTime, ZonedDateTime, LocalDate, LocalDateTime, LocalTime, Date
```

In above types, there are some types cause conversion problem when store data to db, both implemented in com.rethinkdb.ast.Util::toReqlAst()
```
enum:
   No error, but stored as null.
```
```
Integer, int:
  Strange unreasonable exception complains should use Long instead of Integer.
com.rethinkdb.gen.exc.ReqlDriverError: Can't convert com.rethinkdb.TestPojo@79b06cab to a ReqlAst: Make getIntegerProperty of com.rethinkdb.TestPojo@79b06cab Long instead of Integer

	at com.rethinkdb.ast.Util.toMap(Util.java:159)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:120)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:33)
	at com.rethinkdb.model.Arguments.coerceAndAdd(Arguments.java:43)
	at com.rethinkdb.gen.ast.Table.insert(Table.java:45)
	at com.rethinkdb.RethinkDBTest.testTableSelectOfPojoCursor(RethinkDBTest.java:293)
```
```
LocalDateTime:
  Error when try to format non-existence timezone. Error message is weird, seems non sense.
java.time.temporal.UnsupportedTemporalTypeException: Unsupported field: OffsetSeconds

	at java.time.LocalDate.get0(LocalDate.java:680)
	at java.time.LocalDate.getLong(LocalDate.java:659)
	at java.time.LocalDateTime.getLong(LocalDateTime.java:720)
	at java.time.format.DateTimePrintContext$1.getLong(DateTimePrintContext.java:205)
	at java.time.format.DateTimePrintContext.getValue(DateTimePrintContext.java:298)
	at java.time.format.DateTimeFormatterBuilder$OffsetIdPrinterParser.format(DateTimeFormatterBuilder.java:3338)
	at java.time.format.DateTimeFormatterBuilder$CompositePrinterParser.format(DateTimeFormatterBuilder.java:2182)
	at java.time.format.DateTimeFormatter.formatTo(DateTimeFormatter.java:1744)
	at java.time.format.DateTimeFormatter.format(DateTimeFormatter.java:1718)
	at java.time.LocalDateTime.format(LocalDateTime.java:1746)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:90)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:33)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:76)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:33)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:120)
	at com.rethinkdb.ast.Util.toReqlAst(Util.java:33)
	at com.rethinkdb.model.Arguments.coerceAndAdd(Arguments.java:43)
	at com.rethinkdb.gen.ast.Table.insert(Table.java:45)
	at com.rethinkdb.RethinkDBTest.testTableSelectOfPojoCursor(RethinkDBTest.java:293)  
```

After stored data into db, there are conversion problem occurs when retrieve data and convert to original type. The conversion logic is very simple:
```
valueClass.cast(value)
```
which means **strict conversion** by 
```
(SomeClass)value
```
So all errors are same java.lang.ClassCastException: 
```
Enum: Cannot cast java.lang.String to com.rethinkdb.TestPojo$PojoEnum
BigDecimal:  Cannot cast java.lang.Double to java.math.BigDecimal
BigInteger: Cannot cast java.lang.Long to java.math.BigInteger
double: Cannot cast java.lang.Double to double
Float: Cannot cast java.lang.Double to java.lang.Float
float: Cannot cast java.lang.Double to float
Long: Cannot cast java.lang.Double to java.lang.Long
long: Cannot cast java.lang.Long to long
Integer: Cannot cast java.lang.Long to java.lang.Integer
int: Cannot cast java.lang.Long to int
boolean: Cannot cast java.lang.Boolean to boolean
Short: Cannot cast java.lang.Long to java.lang.Short
short: Cannot cast java.lang.Long to short
Byte: Cannot cast java.lang.Long to java.lang.Byte
byte: Cannot cast java.lang.Long to byte
LocalDateTime: Cannot cast java.time.OffsetDateTime to java.time.LocalDateTime
ZonedDateTime: Cannot cast java.time.OffsetDateTime to java.time.ZonedDateTime

Stack trace are same:
java.lang.ClassCastException: Cannot cast yyy to XXXType

	at java.lang.Class.cast(Class.java:3369)
	at com.rethinkdb.net.Util.constructViaPublicParameterlessConstructor(Util.java:115)
	at com.rethinkdb.net.Util.toPojo(Util.java:70)
	at com.rethinkdb.net.Util.convertToPojo(Util.java:39)
	at com.rethinkdb.net.Cursor$DefaultCursor.getNext(Cursor.java:198)
	at com.rethinkdb.net.Cursor.next(Cursor.java:124)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at com.rethinkdb.net.Cursor.toList(Cursor.java:144)
	at com.rethinkdb.RethinkDBTest.testTableSelectOfPojoCursor(RethinkDBTest.java:299)
```
```
LocalDate:

com.rethinkdb.gen.exc.ReqlDriverError: Can't convert {dayOfWeek=SUNDAY, month=MARCH, dayOfMonth=6, dayOfYear=66, era=CE, year=2016, monthValue=3, chronology={calendarType=iso8601, id=ISO}, leapYear=true} to a POJO: class java.time.LocalDate should have a public parameterless constructor or a public constructor with 9 parameters
```
```
LocalTime:

com.rethinkdb.gen.exc.ReqlDriverError: Can't convert {hour=13, nano=24000000, minute=12, second=54} to a POJO: class java.time.LocalTime should have a public parameterless constructor or a public constructor with 4 parameters
```
stack trace are same:
```
	at com.rethinkdb.net.Util.toPojo(Util.java:87)
	at com.rethinkdb.net.Util.constructViaPublicParameterlessConstructor(Util.java:118)
	at com.rethinkdb.net.Util.toPojo(Util.java:74)
	at com.rethinkdb.net.Util.convertToPojo(Util.java:39)
	at com.rethinkdb.net.Cursor$DefaultCursor.getNext(Cursor.java:198)
	at com.rethinkdb.net.Cursor.next(Cursor.java:124)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at com.rethinkdb.net.Cursor.toList(Cursor.java:144)
	at com.rethinkdb.RethinkDBTest.testTableSelectOfPojoCursor(RethinkDBTest.java:299)
```

### These modification does ONLY modify test code, NO any harm, so **i strongly recommend merge this as soon as possible**.

By running the test case <a href="https://github.com/sjitech/rethinkdb/blob/next/drivers/java/src/test/java/com/rethinkdb/RethinkDBTest.java#L290"> testTableSelectOfPojo()</a>, you can see many conversion problems.

### Sorry, if you are stuck in the test case, you need set some problematic type's default value to null in <a href="https://github.com/sjitech/rethinkdb/blob/next/drivers/java/src/test/java/com/rethinkdb/TestPojo.java">TestPojo.java</a>  bean, so you can pass some error and continue test other types.
```
public class TestPojo {
    ...
    private PojoEnum enumProperty = PojoEnum.AAA;
    private OffsetDateTime offsetDateTimeProperty = sampleOffsetDateTimeProperty;
    private LocalDateTime localDateTimeProperty = sampleLocalDateTimeProperty;
    private ZonedDateTime zonedDateTimeProperty = sampleZonedDateTimeProperty;
    private LocalDate localDateProperty = sampleLocalDateProperty;
    private LocalTime localTimeProperty = sampleLocalTimeProperty;
    private Double doubleProperty = Double.MAX_VALUE;
    ... ...
```

Related: https://github.com/rethinkdb/rethinkdb/pull/5483, https://github.com/rethinkdb/rethinkdb/issues/5228, https://github.com/rethinkdb/rethinkdb/issues/5446

### CODE REVIEW

**Comment by QianJin2013** [comment]
I have modified first comment to add more detail description

**Comment by QianJin2013** [comment]
Test result:
```
... ...
19:54:40.828 [main] DEBUG com.rethinkdb.ast.Query - JSON Send: Token: 1 [1,[57,["javatests"]]]
19:54:40.845 [pool-1-thread-1] DEBUG com.rethinkdb.ast.Query - JSON Recv: Token: 1 {"t":1,"r":[{"config_changes":[{"new_val":{"id":"db39251f-dd73-44b9-8347-1b6fc8b1c4ab","name":"javatests"},"old_val":null}],"dbs_created":1}]}
19:54:40.882 [main] DEBUG com.rethinkdb.ast.Query - JSON Send: Token: 2 [1,[177,[[14,["javatests"]]]]]
19:54:40.883 [pool-1-thread-1] DEBUG com.rethinkdb.ast.Query - JSON Recv: Token: 2 {"t":1,"r":[{"ready":0}]}
19:54:40.885 [main] DEBUG com.rethinkdb.ast.Query - JSON Send: Token: 3 [1,[60,[[14,["javatests"]],"atest"]]]
19:54:44.384 [pool-1-thread-1] DEBUG com.rethinkdb.ast.Query - JSON Recv: Token: 3 {"t":1,"r":[{"config_changes":[{"new_val":{"db":"javatests","durability":"hard","id":"f6a96348-e78c-43ca-80ff-e05d0e39b520","indexes":[],"name":"atest","primary_key":"id","shards":[{"nonvoting_replicas":[],"primary_replica":"p14010112_no_MacBook_Pro_local_653","replicas":["p14010112_no_MacBook_Pro_local_653"]}],"write_acks":"majority"},"old_val":null}],"tables_created":1}]}
19:54:44.387 [main] DEBUG com.rethinkdb.ast.Query - JSON Send: Token: 4 [1,[177,[[15,[[14,["javatests"]],"atest"]]]]]
19:54:44.404 [pool-1-thread-1] DEBUG com.rethinkdb.ast.Query - JSON Recv: Token: 4 {"t":1,"r":[{"ready":1}]}19:54:44.415 [main] DEBUG com.rethinkdb.ast.Query - JSON Send: Token: 1 [1,[54,[[15,[[14,["javatests"]],"atest"]]]]]
19:54:44.419 [pool-2-thread-1] DEBUG com.rethinkdb.ast.Query - JSON Recv: Token: 1 {"t":1,"r":[{"deleted":0,"errors":0,"inserted":0,"replaced":0,"skipped":0,"unchanged":0}]}
19:54:44.500 [main] DEBUG com.rethinkdb.ast.Query - JSON Send: Token: 2 [1,[56,[[15,[[14,["javatests"]],"atest"]],[3,[],{"enumProperty":"AAA","localDateTimeProperty":[99,["2016-03-05T19:54:44.444+09"]],"primitiveByteProperty":127,"primitiveShortProperty":32767,"offsetDateTimeProperty":[99,["2016-03-05T19:54:44.443+09"]],"booleanProperty":true,"primitiveFloatProperty":3.4028235E38,"doubleProperty":1.7976931348623157E308,"dateProperty":[3,[],{"date":5,"hours":19,"seconds":44,"month":2,"timezoneOffset":-540,"year":116,"minutes":54,"time":1457175284446,"day":6}],"localTimeProperty":[3,[],{"hour":19,"nano":444000000,"minute":54,"second":44}],"floatProperty":3.4028235E38,"pojoProperty":[3,[],{"booleanProperty":true,"longProperty":42}],"bigDecimalProperty":123456789012345678901234567,"primitiveBooleanProperty":true,"primitiveLongProperty":9223372036854775807,"longProperty":9223372036854775807,"integerProperty":2147483647,"stringProperty":"foo","zonedDateTimeProperty":[99,["2016-03-05T19:54:44.444+09"]],"bigIntegerProperty":123456789012345,"byteProperty":127,"localDateProperty":[3,[],{"dayOfWeek":"SATURDAY","month":"MARCH","dayOfMonth":5,"dayOfYear":65,"era":"CE","year":2016,"monthValue":3,"chronology":[3,[],{"calendarType":"iso8601","id":"ISO"}],"leapYear":true}],"primitiveDoubleProperty":1.7976931348623157E308,"primitiveIntegerProperty":2147483647,"shortProperty":32767}]]]]
19:54:44.510 [pool-2-thread-1] DEBUG com.rethinkdb.ast.Query - JSON Recv: Token: 2 {"t":1,"r":[{"deleted":0,"errors":0,"generated_keys":["6fb65752-bb39-412d-8a7f-5e95e6f470ea"],"inserted":1,"replaced":0,"skipped":0,"unchanged":0}]}
19:54:44.514 [main] DEBUG com.rethinkdb.ast.Query - JSON Send: Token: 3 [1,[56,[[15,[[14,["javatests"]],"atest"]],[3,[],{"enumProperty":"AAA","localDateTimeProperty":[99,["2016-03-05T19:54:44.444+09"]],"primitiveByteProperty":127,"primitiveShortProperty":32767,"offsetDateTimeProperty":[99,["2016-03-05T19:54:44.443+09"]],"booleanProperty":true,"primitiveFloatProperty":3.4028235E38,"doubleProperty":1.7976931348623157E308,"dateProperty":[3,[],{"date":5,"hours":19,"seconds":44,"month":2,"timezoneOffset":-540,"year":116,"minutes":54,"time":1457175284446,"day":6}],"localTimeProperty":[3,[],{"hour":19,"nano":444000000,"minute":54,"second":44}],"floatProperty":3.4028235E38,"pojoProperty":[3,[],{"booleanProperty":false,"longProperty":53}],"bigDecimalProperty":123456789012345678901234567,"primitiveBooleanProperty":true,"primitiveLongProperty":9223372036854775807,"longProperty":9223372036854775807,"integerProperty":2147483647,"stringProperty":"bar","zonedDateTimeProperty":[99,["2016-03-05T19:54:44.444+09"]],"bigIntegerProperty":123456789012345,"byteProperty":127,"localDateProperty":[3,[],{"dayOfWeek":"SATURDAY","month":"MARCH","dayOfMonth":5,"dayOfYear":65,"era":"CE","year":2016,"monthValue":3,"chronology":[3,[],{"calendarType":"iso8601","id":"ISO"}],"leapYear":true}],"primitiveDoubleProperty":1.7976931348623157E308,"primitiveIntegerProperty":2147483647,"shortProperty":32767}]]]]
19:54:44.525 [pool-2-thread-1] DEBUG com.rethinkdb.ast.Query - JSON Recv: Token: 3 {"t":1,"r":[{"deleted":0,"errors":0,"generated_keys":["4d0099eb-f0d4-4e98-94fc-15b47f67e8d2"],"inserted":1,"replaced":0,"skipped":0,"unchanged":0}]}
19:54:44.525 [main] DEBUG com.rethinkdb.ast.Query - JSON Send: Token: 4 [1,[15,[[14,["javatests"]],"atest"]]]
19:54:44.527 [pool-2-thread-1] DEBUG com.rethinkdb.ast.Query - JSON Recv: Token: 4 {"t":2,"r":[{"bigDecimalProperty":1.2345678901234568e26,"bigIntegerProperty":123456789012345,"booleanProperty":true,"byteProperty":127,"dateProperty":{"date":5,"day":6,"hours":19,"minutes":54,"month":2,"seconds":44,"time":1457175284446,"timezoneOffset":-540,"year":116},"doubleProperty":1.7976931348623157e308,"enumProperty":"AAA","floatProperty":3.4028235e38,"id":"4d0099eb-f0d4-4e98-94fc-15b47f67e8d2","integerProperty":2147483647,"localDateProperty":{"chronology":{"calendarType":"iso8601","id":"ISO"},"dayOfMonth":5,"dayOfWeek":"SATURDAY","dayOfYear":65,"era":"CE","leapYear":true,"month":"MARCH","monthValue":3,"year":2016},"localDateTimeProperty":{"$reql_type$":"TIME","epoch_time":1457175284.444,"timezone":"+09:00"},"localTimeProperty":{"hour":19,"minute":54,"nano":444000000,"second":44},"longProperty":9223372036854776000.0,"offsetDateTimeProperty":{"$reql_type$":"TIME","epoch_time":1457175284.443,"timezone":"+09:00"},"pojoProperty":{"booleanProperty":false,"longProperty":53},"primitiveBooleanProperty":true,"primitiveByteProperty":127,"primitiveDoubleProperty":1.7976931348623157e308,"primitiveFloatProperty":3.4028235e38,"primitiveIntegerProperty":2147483647,"primitiveLongProperty":9223372036854776000.0,"primitiveShortProperty":32767,"shortProperty":32767,"stringProperty":"bar","zonedDateTimeProperty":{"$reql_type$":"TIME","epoch_time":1457175284.444,"timezone":"+09:00"}},{"bigDecimalProperty":1.2345678901234568e26,"bigIntegerProperty":123456789012345,"booleanProperty":true,"byteProperty":127,"dateProperty":{"date":5,"day":6,"hours":19,"minutes":54,"month":2,"seconds":44,"time":1457175284446,"timezoneOffset":-540,"year":116},"doubleProperty":1.7976931348623157e308,"enumProperty":"AAA","floatProperty":3.4028235e38,"id":"6fb65752-bb39-412d-8a7f-5e95e6f470ea","integerProperty":2147483647,"localDateProperty":{"chronology":{"calendarType":"iso8601","id":"ISO"},"dayOfMonth":5,"dayOfWeek":"SATURDAY","dayOfYear":65,"era":"CE","leapYear":true,"month":"MARCH","monthValue":3,"year":2016},"localDateTimeProperty":{"$reql_type$":"TIME","epoch_time":1457175284.444,"timezone":"+09:00"},"localTimeProperty":{"hour":19,"minute":54,"nano":444000000,"second":44},"longProperty":9223372036854776000.0,"offsetDateTimeProperty":{"$reql_type$":"TIME","epoch_time":1457175284.443,"timezone":"+09:00"},"pojoProperty":{"booleanProperty":true,"longProperty":42},"primitiveBooleanProperty":true,"primitiveByteProperty":127,"primitiveDoubleProperty":1.7976931348623157e308,"primitiveFloatProperty":3.4028235e38,"primitiveIntegerProperty":2147483647,"primitiveLongProperty":9223372036854776000.0,"primitiveShortProperty":32767,"shortProperty":32767,"stringProperty":"foo","zonedDateTimeProperty":{"$reql_type$":"TIME","epoch_time":1457175284.444,"timezone":"+09:00"}}],"n":[]}
19:54:44.538 [main] DEBUG com.rethinkdb.ast.Query - JSON Send: Token: 4 [2]
19:54:44.539 [pool-2-thread-1] DEBUG com.rethinkdb.ast.Query - JSON Recv: Token: 4 {"t":16,"r":["Token 4 not in stream cache."],"b":[]}
19:54:44.563 [main] DEBUG com.rethinkdb.ast.Query - JSON Send: Token: 1 [1,[61,[[14,["javatests"]],"atest"]]]
19:54:44.578 [pool-3-thread-1] DEBUG com.rethinkdb.ast.Query - JSON Recv: Token: 1 {"t":1,"r":[{"config_changes":[{"new_val":null,"old_val":{"db":"javatests","durability":"hard","id":"f6a96348-e78c-43ca-80ff-e05d0e39b520","indexes":[],"name":"atest","primary_key":"id","shards":[{"nonvoting_replicas":[],"primary_replica":"p14010112_no_MacBook_Pro_local_653","replicas":["p14010112_no_MacBook_Pro_local_653"]}],"write_acks":"majority"}}],"tables_dropped":1}]}
19:54:44.581 [main] DEBUG com.rethinkdb.ast.Query - JSON Send: Token: 2 [1,[58,["javatests"]]]
19:54:44.583 [pool-3-thread-1] DEBUG com.rethinkdb.ast.Query - JSON Recv: Token: 2 {"t":1,"r":[{"config_changes":[{"new_val":null,"old_val":{"id":"db39251f-dd73-44b9-8347-1b6fc8b1c4ab","name":"javatests"}}],"dbs_dropped":1,"tables_dropped":0}]}
Disconnected from the target VM, address: '127.0.0.1:64761', transport: 'socket'

Process finished with exit code 0

```

**Comment by QianJin2013** [comment]
For more easy to understand how data is stored in db, i paste beautified result of raw selected result (JSON), now all of these types works well.
```
{
    "enumProperty": "AAA",

    "offsetDateTimeProperty": {
        "$reql_type$": "TIME",
        "epoch_time": 1457175284.443,
        "timezone": "+09:00"
    },
    "localDateTimeProperty": {
        "$reql_type$": "TIME",
        "epoch_time": 1457175284.444,
        "timezone": "+09:00"
    },
    "localTimeProperty": {
        "hour": 19,
        "minute": 54,
        "nano": 444000000,
        "second": 44
    },
    "zonedDateTimeProperty": {
        "$reql_type$": "TIME",
        "epoch_time": 1457175284.444,
        "timezone": "+09:00"
    },

    "localDateProperty": {
        "chronology": {
            "calendarType": "iso8601",
            "id": "ISO"
        },
        "dayOfMonth": 5,
        "dayOfWeek": "SATURDAY",
        "dayOfYear": 65,
        "era": "CE",
        "leapYear": true,
        "month": "MARCH",
        "monthValue": 3,
        "year": 2016
    },

    "dateProperty": {
        "date": 5,
        "day": 6,
        "hours": 19,
        "minutes": 54,
        "month": 2,
        "seconds": 44,
        "time": 1457175284446,
        "timezoneOffset": -540,
        "year": 116
    },

    "bigDecimalProperty": 1.2345678901234568e+26,
    "bigIntegerProperty": 123456789012345,

    "doubleProperty": 1.7976931348623157e+308,
    "primitiveDoubleProperty": 1.7976931348623157e+308,

    "floatProperty": 3.4028235e+38,
    "primitiveFloatProperty": 3.4028235e+38,

    "booleanProperty": true,
    "primitiveBooleanProperty": true,

    "longProperty": 9.223372036854776e+18,
    "primitiveLongProperty": 9.223372036854776e+18,

    "integerProperty": 2147483647,
    "primitiveIntegerProperty": 2147483647,

    "shortProperty": 32767,
    "primitiveShortProperty": 32767,

    "byteProperty": 127,
    "primitiveByteProperty": 127,

    "stringProperty": "foo",

    "pojoProperty": {
        "booleanProperty": true,
        "longProperty": 42
    },

    "id": "6fb65752-bb39-412d-8a7f-5e95e6f470ea"
}
```

**Comment by deontologician** [comment]
This looks awesome, thank you @QianJin2013 . I am looking forward to reviewing it Monday :)

**Comment by QianJin2013** [comment]
Sorry, i have updated first comment again to add some description of db -> java conversion error.

File: drivers/java/src/test/java/com/rethinkdb/RethinkDBTest.java [hunk 1]
```diff
@@ -488,6 +486,85 @@ public void testSaveObjectIgnoreNullPropertiesNested() {
         assertEquals(pojoOneInnerSelected.get("notNullProperty"), pojoOneInner.getNotNullProperty());
     }
 
+    @Test
+    public void testTableSelect_convertStringToDate() {
+        Locale.setDefault(Locale.JAPAN);
+        _convertStringToDate("2016-03-16", new String[]{
```
Comment (QianJin2013) [inline drivers/java/src/test/java/com/rethinkdb/RethinkDBTest.java]: When convert db string to `Date`/`OffsetDateTime`/`LocalDateTime`/`LocalDate`,
Support following common used date format:

```
2016-03-16, 
2016-03-16T09:58:59, 
2016-03-16T09:58:59.007, 
2016-03-16T09:58:59.007+09:00, 
Sat Dec 30 09:58:59 JST 2016
```

Let these string can be converted to any of above date types.

**Pull request closed without merging**